File: BhaktiChatAndroidNative/app/src/main/java/com/bhaktichat/app/util/VoiceAudioPlayer.kt
package com.bhaktichat.app.util

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.os.Process
import android.util.Log
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * Plays the guide's spoken reply for Voice Mode — continuously-arriving PCM16 mono @ 24kHz
 * chunks from OpenAI's Realtime API (`response.output_audio.delta` events), not a discrete
 * audio file. Runs its own playback thread so the WebSocket's message-handling thread is
 * never blocked by [AudioTrack.write].
 *
 * The queue is deliberately UNBOUNDED: the server sends a reply's audio in a burst far
 * faster than realtime playback drains it. The previous bounded queue (64 chunks) silently
 * dropped whatever didn't fit — heard as stuttery/"slow" speech that cut off mid-sentence
 * on real networks. iOS's AVAudioPlayerNode buffers unboundedly, which is why it never had
 * this bug; a full minute of PCM16@24kHz is only ~2.9MB, so memory is a non-issue.
 */
class VoiceAudioPlayer {
    private var audioTrack: AudioTrack? = null
    private var playbackThread: Thread? = null
    private val isRunning = AtomicBoolean(false)
    private val queue = LinkedBlockingQueue<ByteArray>()

    // Frames written to the track since the last flush; compared against the actual playback
    // head so we can tell when the guide has *finished being heard*, not just finished
    // generating (generation completes ~1-2s before the buffered audio finishes playing).
    private val framesWritten = AtomicLong(0)
    private val generationComplete = AtomicBoolean(false)

    /** Fired on the playback thread once the last generated audio has actually been played out. */
    var onGuideFinishedSpeaking: (() -> Unit)? = null

    fun start() {
        if (isRunning.get()) return

        val minBuffer = AudioTrack.getMinBufferSize(SAMPLE_RATE, CHANNEL_OUT, ENCODING)
        val track = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setSampleRate(SAMPLE_RATE)
                    .setChannelMask(CHANNEL_OUT)
                    .setEncoding(ENCODING)
                    .build()
            )
            // Generous buffer to ride out scheduling jitter — underruns here are audible as
            // stutter. Barge-in latency is unaffected: interruptNow() pause+flushes.
            .setBufferSizeInBytes(minBuffer * 4)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()
        audioTrack = track

        isRunning.set(true)
        playbackThread = Thread({
            Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
            track.play()
            var loggedFirstAudio = false
            while (isRunning.get()) {
                val bytes = queue.poll(20, TimeUnit.MILLISECONDS)
                if (bytes != null) {
                    if (!loggedFirstAudio) {
                        loggedFirstAudio = true
                        Log.i(TAG, "First guide audio written to speaker (${bytes.size} bytes).")
                    }
                    track.write(bytes, 0, bytes.size, AudioTrack.WRITE_BLOCKING)
                    framesWritten.addAndGet((bytes.size / 2).toLong()) // PCM16 mono: 2 bytes/frame
                } else if (generationComplete.get()) {
                    // Queue drained and the model is done generating — has the audio actually
                    // been *heard* yet? The playback head lags what we've written by the track's
                    // internal buffer. Only signal "done speaking" once it catches up.
                    val head = track.playbackHeadPosition.toLong() and 0xFFFFFFFFL
                    if (head >= framesWritten.get()) {
                        generationComplete.set(false)
                        Log.i(TAG, "Guide finished speaking (underruns=${track.underrunCount})")
                        onGuideFinishedSpeaking?.invoke()
                    }
                }
            }
        }, "voice-playback").apply { start() }
    }

    /** Queues a chunk for playback. Never blocks and never drops — safe from any thread. */
    fun enqueue(bytes: ByteArray) {
        queue.offer(bytes)
    }

    /** Signals that no more audio will arrive for the current reply, so the playback thread can
     *  watch for the moment the buffered audio finishes playing and fire [onGuideFinishedSpeaking]. */
    fun markGenerationComplete() {
        generationComplete.set(true)
    }

    /**
     * Stops playback immediately (barge-in) without tearing down the track — call this the
     * moment the user starts talking over the guide, optimistically, before waiting for the
     * server's own interruption confirmation event.
     */
    fun interruptNow() {
        val track = audioTrack ?: return
        queue.clear()
        if (track.playState == AudioTrack.PLAYSTATE_PLAYING) {
            track.pause()
            track.flush()
        }
        // flush() resets the playback head to 0, so the written-frame counter must reset too,
        // and a barge-in must not later fire a stale "finished speaking" for the killed reply.
        framesWritten.set(0)
        generationComplete.set(false)
        track.play()
    }

    fun stop() {
        if (!isRunning.compareAndSet(true, false)) return
        generationComplete.set(false)
        framesWritten.set(0)
        queue.clear()
        playbackThread?.join(200)
        audioTrack?.let { track ->
            if (track.playState == AudioTrack.PLAYSTATE_PLAYING) track.pause()
            track.flush()
            track.stop()
            track.release()
        }
        audioTrack = null
        playbackThread = null
    }

    companion object {
        private const val TAG = "VoiceAudioPlayer"
        const val SAMPLE_RATE = 24_000
        private const val CHANNEL_OUT = AudioFormat.CHANNEL_OUT_MONO
        private const val ENCODING = AudioFormat.ENCODING_PCM_16BIT
    }
}

File: BhaktiChatiOS/Sources/Core/Services/VoiceAudioPlayer.swift
import Foundation
#if os(iOS)
import AVFoundation

/// Plays the guide's spoken reply for Voice Mode — continuously-arriving PCM16 mono @ 24kHz
/// chunks from OpenAI's Realtime API (`response.output_audio.delta` events). Uses
/// `AVAudioEngine` + `AVAudioPlayerNode`; each scheduled buffer's `.dataPlayedBack`
/// completion fires once that buffer has actually been rendered to the speaker — not merely
/// consumed off the queue — so `onGuideFinishedSpeaking` only fires once the guide has truly
/// been heard, not the moment the model finishes generating. Mirrors Android's
/// `util/VoiceAudioPlayer.kt` (which achieves the same thing via `AudioTrack`'s playback-head
/// position).
final class VoiceAudioPlayer {
    private let engine = AVAudioEngine()
    private let playerNode = AVAudioPlayerNode()
    private var isRunning = false

    // Bumped on every interrupt/stop so a completion handler from a buffer that was
    // superseded by a barge-in can't fire a stale "finished speaking" signal.
    private var generation = 0
    private var pendingBufferCount = 0
    private var isGenerationComplete = false

    var onGuideFinishedSpeaking: (() -> Void)?

    private static let format = VoiceAudioCapture.targetFormat

    func start() {
        guard !isRunning else { return }
        engine.attach(playerNode)
        engine.connect(playerNode, to: engine.mainMixerNode, format: Self.format)
        engine.prepare()
        do {
            try engine.start()
            playerNode.play()
            isRunning = true
        } catch {
            // isRunning stays false; enqueue() becomes a no-op below.
        }
    }

    /// Queues a chunk for playback. Safe to call from any thread — completion handling
    /// always hops back to main before touching shared state.
    func enqueue(_ data: Data) {
        guard isRunning else { return }
        let frameCount = UInt32(data.count / MemoryLayout<Int16>.size)
        guard frameCount > 0,
              let buffer = AVAudioPCMBuffer(pcmFormat: Self.format, frameCapacity: frameCount) else { return }
        buffer.frameLength = frameCount
        data.withUnsafeBytes { raw in
            guard let source = raw.bindMemory(to: Int16.self).baseAddress,
                  let dest = buffer.int16ChannelData?[0] else { return }
            dest.update(from: source, count: Int(frameCount))
        }

        let bufferGeneration = generation
        pendingBufferCount += 1
        playerNode.scheduleBuffer(buffer, completionCallbackType: .dataPlayedBack) { [weak self] _ in
            DispatchQueue.main.async {
                guard let self, bufferGeneration == self.generation else { return }
                self.pendingBufferCount = max(0, self.pendingBufferCount - 1)
                if self.pendingBufferCount == 0 && self.isGenerationComplete {
                    self.isGenerationComplete = false
                    self.onGuideFinishedSpeaking?()
                }
            }
        }
    }

    /// Signals that no more audio will arrive for the current reply. Once every already-
    /// scheduled buffer has actually finished playing, `onGuideFinishedSpeaking` fires.
    func markGenerationComplete() {
        isGenerationComplete = true
        if pendingBufferCount == 0 {
            isGenerationComplete = false
            onGuideFinishedSpeaking?()
        }
    }

    /// Stops playback immediately (barge-in) — call the moment the user starts talking over
    /// the guide, optimistically, before waiting for the server's own interruption event.
    func interruptNow() {
        generation += 1
        pendingBufferCount = 0
        isGenerationComplete = false
        playerNode.stop()
        playerNode.play()
    }

    func stop() {
        guard isRunning else { return }
        isRunning = false
        generation += 1
        pendingBufferCount = 0
        isGenerationComplete = false
        playerNode.stop()
        engine.stop()
    }
}
#else
final class VoiceAudioPlayer {
    var onGuideFinishedSpeaking: (() -> Void)?
    func start() {}
    func enqueue(_ data: Data) {}
    func markGenerationComplete() {}
    func interruptNow() {}
    func stop() {}
}
#endif
